### PR TITLE
feat: useRequest 支持自定义缓存工具 #875

### PR DESCRIPTION
**Pull request by izhangchao**
[[中文版模板 / Chinese template](https://github.com/alibaba/hooks/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

File: packages/hooks/src/useRequest/doc/cache/demo/customCache.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            });
+        }, 1000);
+    });
+}
+
+const Article = () => {
+    const { data, loading } = useRequest(getArticle, {
+        cacheKey: 'cacheKey-demo',
+        getCache: (cacheKey) => {
+            console.log('AAAAAAAAAAA', cacheKey)
```
Comment (brickspert) [inline packages/hooks/src/useRequest/doc/cache/demo/customCache.tsx]: remove

File: packages/hooks/src/useRequest/doc/cache/demo/customCache.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { useBoolean } from 'ahooks';
+import Mock from 'mockjs';
+import React from 'react';
+import { useRequest } from 'ahooks';
+
+async function getArticle(): Promise<{ data: string; time: number }> {
+    console.log('cacheKey');
```
Comment (brickspert) [inline packages/hooks/src/useRequest/doc/cache/demo/customCache.tsx]: remove

File: packages/hooks/src/useRequest/doc/cache/cache.zh-CN.md [hunk 3]
```diff
@@ -46,6 +46,13 @@ group:
 
 <code src="./demo/params.tsx" />
 
+
+### 自定义缓存
```
Comment (brickspert) [inline packages/hooks/src/useRequest/doc/cache/cache.zh-CN.md:50]: 英文文档需要同时更新

File: packages/hooks/src/useRequest/src/plugins/useCachePlugin.ts [hunk 4]
```diff
@@ -6,104 +6,115 @@ import * as cache from '../utils/cache';
...
-  fetchInstance,
-  {
-    cacheKey,
-    cacheTime = 5 * 60 * 1000,
-    staleTime = 0,
-    // cacheType,
-    // setCache,
-    // getCache,
-  },
+    fetchInstance,
```
Comment (brickspert) [inline packages/hooks/src/useRequest/src/plugins/useCachePlugin.ts:10]: 不要替换前面的空格~

File: packages/hooks/src/useRequest/src/plugins/useCachePlugin.ts [hunk 5]
```diff
@@ -6,104 +6,115 @@ import * as cache from '../utils/cache';
...
+    const currentPromiseRef = useRef<Promise<any>>();
 
-  useCreation(() => {
-    if (!cacheKey) {
-      return;
-    }
+    // get|set data from custom cache when seted in options
+    const customCache = useCreation(() => ({
+        setCache: (cacheKey, cacheTime, data, params) => {
+            return !!setCache ? setCache(cacheKey, data) : cache.setCache(cacheKey, cacheTime, data, params);
```
Comment (brickspert) [inline packages/hooks/src/useRequest/src/plugins/useCachePlugin.ts]: 这里不建议用 useCreation，需要考虑 setCache 的变化。

另外就是 setCache 要不要缓存 cacheTime 和 params。

**Review by brickspert — CHANGES_REQUESTED** [review]

**Comment by izhangchao** [comment]
@brickspert

**Comment by baranwang** [comment]
正需要这个功能，刚想自己PR一个看了一眼就看到这个PR

@izhangchao 你的代码缩进是四个空格吧，跟项目不一致，所以没合并（我猜是这个原因

https://github.com/alibaba/hooks/blob/78f694fd5f93a338947be612bf7448254b5a1a2c/.editorconfig#L6

**Comment by brickspert** [comment]
这个 feature 我可能需要大改一下。

**Pull request closed without merging**